dags/glam.py

```diff
@@ -0,0 +1,147 @@
+from airflow import DAG
+from datetime import datetime, timedelta
+from utils.gcp import bigquery_etl_query
+from airflow.operators.sensors import ExternalTaskSensor
+
+
+dataset_id = "telemetry_derived"
+default_args = {
+    'owner': 'msamuel@mozilla.com',
+    'depends_on_past': False,
+    'start_date': datetime(2019, 10, 22),
+    'email': ['telemetry-alerts@mozilla.com', 'msamuel@mozilla.com'],
+    'email_on_failure': True,
+    'email_on_retry': True,
+    'retries': 1,
+    'retry_delay': timedelta(minutes=30),
+}
+
+dag = DAG('glam', default_args=default_args, schedule_interval='@daily')
+
+# Make sure all the data for the given day has arrived before running.
+wait_for_main_ping = ExternalTaskSensor(
+    task_id="wait_for_main_ping",
+    external_dag_id="main_summary",
+    external_task_id="copy_deduplicate_main_ping",
+    execution_delta=timedelta(hours=-1),
+    check_existence=True,
+    dag=dag,
+)
+
+clients_daily_scalar_aggregates = bigquery_etl_query(
+    task_id="clients_daily_scalar_aggregates",
+    destination_table="clients_daily_scalar_aggregates_v1",
+    dataset_id=dataset_id,
+    owner="msamuel@mozilla.com",
+    email=["telemetry-alerts@mozilla.com", "msamuel@mozilla.com"],
+    depends_on_past=False,
+    dag=dag)
+
+sql_file_path = "sql/{}/{}/query.sql".format(dataset_id, "clients_daily_keyed_scalar_aggregates_v1")
+clients_daily_keyed_scalar_aggregates = bigquery_etl_query(
+    task_id="clients_daily_keyed_scalar_aggregates",
+    destination_table="clients_daily_scalar_aggregates_v1",
```

Comment:
It looks like the destination tables in a few of these queries don't match the names, so that's probably an error.

Also, because Airflow now has privs to issue query jobs in shared-prod, it shouldn't be necessary to specify the sql_file_path explicitly. If you set project, dataset, and destination_table, it should be able to infer the correct sql_file_path.

Reply:
Actually this was intentional because there are 3 queries that end up into 1 table:
`clients_daily_scalar_aggregates_v1` + `clients_daily_keyed_scalar_aggregates_v1` + `clients_daily_keyed_boolean_aggregates_v1` -> `clients_daily_scalar_aggregates_v1`

And similarly 2 queries go into this other table:
`clients_daily_histogram_aggregates_v1` + `clients_daily_keyed_histogram_aggregates_v1` -> `clients_daily_histogram_aggregates_v1`

Now that I write this out, I realize the naming can be a tad misleading... 
but generally it's because keyed scalars and keyed booleans are processed in a slightly different way than regular scalars so they need their own query but the end result of the aggregation is about the same so it can be included in the same table. Same idea for histograms.

Reply:
Fascinating. And I see now that you set `--noreplace` so that these all append.

A concern I have with this approach is reprocessing. We'd have to manually delete a day before running this DAG again, or we'd end up with duplicate results. It's a nice feature of the default parameters that we generally have ETL queries replace partitions rather than appending, so that reprocessing is always "safe".

Have you considered moving these three queries into a single query file and using `UNION ALL` between them? Or, we could have these be three separate tables with identical schemas and have a single view on top of them that does the UNION. Do either of those sound like reasonable options?

Reply:
Hm. I think this may not be an issue because `clients_daily_scalar_aggregates_v1` runs first and it doesn't have a `--no-replace` tag so it replaces the data in the partition. The other queries append afterwards. Same for histograms. That is my understanding of how this dag will run.

If I'm missing something about how this works, I think I'd rather go with the second option of 3 tables with a view on top.

Reply:
> Hm. I think this may not be an issue because `clients_daily_scalar_aggregates_v1` runs first and it doesn't have a `--no-replace` tag so it replaces the data in the partition. The other queries append afterwards. Same for histograms. That is my understanding of how this dag will run.

That sounds fine. Thanks for running through that. It's probably worth adding a comment above this section explaining the approach, since it may be difficult to infer it from reading through the individual task configs.


> If I'm missing something about how this works, I think I'd rather go with the second option of 3 tables with a view on top because BigQuery tends to give me `complex query` errors, at least in console, when there are too many parts to a query...

That is indeed a very good reason to keep these separated into different queries 👍